File: codewars/6kyu/two-sum/dduck-two-sum.py
# Two Sum
# https://www.codewars.com/kata/52c31f8e6605bcc646000082

from unittest import TestCase


def two_sum(numbers, target):
    copied_numbers = sorted(filter(lambda x: x <= target, numbers))
    left_index = 0
    right_index = len(copied_numbers) - 1

    while left_index < right_index:
        total = copied_numbers[left_index] + copied_numbers[right_index]
        if total == target:
            break
        if total < target:
            left_index += 1
        else:
            right_index -= 1

    return [numbers.index(copied_numbers[left_index]),
            len(numbers) - 1 - numbers[::-1].index(copied_numbers[right_index])]


TestCase().assertEqual(sorted(two_sum([1, 2, 3], 4)), [0, 2])
TestCase().assertEqual(sorted(two_sum([1234, 5678, 9012], 14690)), [1, 2])
TestCase().assertEqual(sorted(two_sum([2, 2, 3], 4)), [0, 1])

File: codewars/6kyu/two-sum/dduck-two-sum.js
// Two Sum
// https://www.codewars.com/kata/52c31f8e6605bcc646000082

const twoSum = (numbers, target) => {
  let copiedNumbers = numbers.filter((v) => v <= target).sort((a, b) => a - b);
  let leftIndex = 0;
  let rightIndex = copiedNumbers.length - 1;
  let sum = 0;
  
  while (leftIndex < rightIndex) {
    sum = copiedNumbers[leftIndex] + copiedNumbers[rightIndex];
    if (sum === target) {
      break;
    }
    if (sum < target) {
      leftIndex += 1;
    } else {
      rightIndex -= 1;
    }
  }
  
  return [numbers.indexOf(copiedNumbers[leftIndex]), numbers.lastIndexOf(copiedNumbers[rightIndex])];
};

console.log(twoSum([1, 2, 3], 4), [0, 2]);
console.log(twoSum([1234, 5678, 9012], 14690), [1, 2]);
console.log(twoSum([2, 2, 3], 4), [0, 1]);
console.log(twoSum([42, 769, 396, 74, 706, 736, 967, 201, 974, 801], 875));
